js/runner.go

```diff
@@ -372,6 +374,8 @@ type VU struct {
 	// goroutine per call.
 	interruptTrackedCtx context.Context
 	interruptCancel     context.CancelFunc
+
+	m *sync.Mutex
```

Comment:
Can you explain why `mu sync.Mutex` not work 🤔

Reply:
It probably would, but if somewhere in the k6 code we decide on copying the `VU` instead of `*VU` it would stop working and given the state of the `js` package and the amount of race conditions, alone, that I have found so far in it, I prefer if I don't leave it to chance ...
 
Also as mentioned in the commit message this is likely to be a temporary fix as it also doesn't fix every possible way we can run two ... things in single VU

Reply:
Not sure I can think what is the case of `would stop working`. As I understand, if we decide to copy the `VU`, we want they will work independently right? So if you use `m *sync.Mutex`, then that `m` will lock both `VU` at the same time. Using `sync.Mutex`,  both original `Vu` and the copied one will work independently from each others.

Maybe I don't have the full picture of the code, but in any case, can you add a TODO or comment to mention that reason, so we don't forget this in the future.

Reply:
@cuonglm, as the [Go documentation states](https://golang.org/pkg/sync/#Mutex), "A Mutex must not be copied after first use". So we can't just copy it. So I don't mind erring on the side of caution here and using a pointer.... Copies can happen inadvertently, for example by methods with value receivers. See https://eli.thegreenplace.net/2018/beware-of-copying-mutexes-in-go/ for more details

Reply:
@na-- I think the main point is a mutex can lock 2 VU is acceptable or not? The main concern of copying `VU` or `*VU` can be easy done by introduce `VU.Copy` method. We are worrying about a very unlikely problem to happen, and prevent it by using a way which can be harder to maintain and debug code later.

Reply:
@na-- Using a pointer, now mutex can be locked both `VUs`. If one `VU` get the lock, then crash, the second `VU` will be stuck. It tastes not well, two different struct shares the same mutex. For example https://play.golang.org/p/2mg9kkGfI88

Reply:
There's some significant misunderstanding, we almostWe seem to be talking about different things. There is no "both VUs" - we're not cloning the VU, we don't want to clone the VU, there would never be 2 VUs copied from the same struct :smile: One `VU` struct instance will only ever be a single VU. 

See https://eli.thegreenplace.net/2018/beware-of-copying-mutexes-in-go/ again. The whole (potential, future) issue is with us potentially adding a method to the `VU` in the future that would have a value receiver. That is, if the method is `func (u VU) someNewMethod() {}` instead of `func (u *VU) someNewMethod() {}` - then `u` and everything in it (including the mutex) would be copied by value. If we use a pointer, like @MStoykov has done, that would never be an issue, but with a non-pointer mutex, we have to be careful to always, always use pointer receivers. Same concerns apply when we pass `VU` as an argument, assign it to a variable, etc. 

All of these

Reply:
@na-- Ah ok, so it's clear now.